lib/base/Page.ts

```diff
@@ -0,0 +1,278 @@
+  _version: TVersion;
+  _payload: TPayload;
+  _solution: Solution;
+
+  /**
+   * @constructor
+   *
+   * @description Base page object to maintain request state.
+   *
+   * @param {Version} version - A twilio version instance
+   * @param {Object} response - The http response
+   * @param {Object} solution - path solution
+   */
+  constructor(
+    version: TVersion,
+    response: Response<string | TPayload>,
+    solution: Solution
+  ) {
+    let payload = this.processResponse(response);
+
+    this._version = version;
+    this._payload = payload;
+    this._solution = solution;
+
+    this.nextPageUrl = this.getNextPageUrl();
+    this.previousPageUrl = this.getPreviousPageUrl();
+    this.instances = this.loadInstances(this.loadPage(payload));
+  }
+
+  /**
+   * @constant META_KEYS
+   * @description meta keys returned in a list request
+   */
+  static META_KEYS: META_KEYS[] = [
+    "end",
+    "first_page_uri",
+    "last_page_uri",
+    "next_page_uri",
+    "num_pages",
+    "page",
+    "page_size",
+    "previous_page_uri",
+    "start",
+    "total",
+    "uri",
+  ];
+
+  /**
+   * Get the url of the previous page of records
+   *
+   * @return {string|undefined} url of the previous page
+   */
+  getPreviousPageUrl(): string {
+    if (
+      "meta" in this._payload &&
+      "previous_page_url" in this._payload.meta &&
+      this._payload.meta.previous_page_url
+    ) {
+      // jshint ignore:line
+      return this._payload.meta.previous_page_url; // jshint ignore:line
+    }
+
+    if (
+      "previous_page_uri" in this._payload &&
+      this._payload.previous_page_uri
+    ) {
+      // jshint ignore:line
+      return this._version._domain.absoluteUrl(this._payload.previous_page_uri); // jshint ignore:line
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Get the url of the next page of records
+   *
+   * @return {string|undefined} url of the next page
+   */
+  getNextPageUrl(): string {
+    if (
+      "meta" in this._payload &&
+      "next_page_url" in this._payload.meta &&
+      this._payload.meta.next_page_url
+    ) {
+      // jshint ignore:line
+      return this._payload.meta.next_page_url; // jshint ignore:line
+    }
+
+    if ("next_page_uri" in this._payload && this._payload.next_page_uri) {
+      // jshint ignore:line
+      return this._version._domain.absoluteUrl(this._payload.next_page_uri); // jshint ignore:line
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Build a new instance given a json payload
+   * @abstract
+   *
+   * @param {object} payload - Payload response from the API
+   * @return {object} instance of a resource
+   */
+  getInstance(payload: any) {
+    throw new Error(
+      "Page.get_instance() must be implemented in the derived class"
+    );
+  }
+
+  /**
+   * Load a list of records
+   *
+   * @param  {object} resources json payload of records
+   * @return {Array} list of resources
+   */
+  loadInstances(resources: TResource[] | TInstance[]): TInstance[] {
```

Comment:
Could `resources` be just `TResource[]`? `TInstance[]` was not specify in the previous d.ts

Reply:
Done